libsplinter/src/service/scabbard/rest_api.rs

```diff
@@ -185,11 +191,67 @@ pub fn make_get_batch_status_endpoint() -> ServiceEndpoint {
                 );
             };
 
-            if let Ok(status) = scabbard.get_batch_info(ids) {
-                Box::new(HttpResponse::Ok().json(status).into_future())
-            } else {
-                Box::new(HttpResponse::InternalServerError().finish().into_future())
+            let wait = query
+                .get("wait")
+                .and_then(|wait_str| {
+                    if wait_str.as_str() == "false" {
+                        None
+                    } else {
+                        wait_str
+                            .parse()
+                            .ok()
+                            .or(Some(DEFAULT_BATCH_STATUS_WAIT_SECS))
+                    }
+                })
+                .map(Duration::from_secs);
+
+            let timeout = match wait {
+                Some(wait_time) => match Instant::now().checked_add(wait_time) {
+                    Some(t) => Some(t),
+                    None => {
+                        return Box::new(
+                            HttpResponse::InternalServerError()
+                                .json("failed to determine wait time")
+                                .into_future(),
+                        )
+                    }
+                },
+                None => None,
+            };
+
+            match get_statuses(&scabbard, &ids, timeout) {
+                Ok(statuses) => Box::new(HttpResponse::Ok().json(statuses).into_future()),
+                Err(err) => Box::new(HttpResponse::InternalServerError().json(err).into_future()),
             }
         }),
     }
 }
+
+fn get_statuses(
+    scabbard: &Scabbard,
+    ids: &[String],
+    timeout: Option<Instant>,
+) -> Result<Vec<BatchInfo>, String> {
+    let batch_infos = scabbard
+        .get_batch_info(ids)
+        .map_err(|err| err.to_string())?;
+
+    if batch_infos
+        .iter()
+        .any(|info| info.status == BatchStatus::Pending)
+    {
+        if let Some(timeout) = timeout {
+            let now = Instant::now();
+            if now < timeout {
+                let normal_wait = Duration::from_millis(BATCH_STATUS_RETRY_INTERVAL_MILLIS);
+                let time_remaining = timeout - now;
+                sleep(min(normal_wait, time_remaining));
```

Comment:
We should capture a story to remove this sleep.

Reply:
A story has been captured for this